[Core] Optimize encoder cache with mask-based storage

sunYtokki opened the pull request:
## Purpose
Optimizes encoder cache memory usage by storing only actual embeddings instead of full placeholder tensors using `is_embed` masks. This addresses the memory inefficiency identified in https://github.com/vllm-project/vllm/issues/25903

**Key improvements:**
- Stores only embedding positions marked in `is_embed` mask (not full scattered tensors)
- Adds `PlaceholderRange.get_num_embeds()` and `extract_embeds_range()` methods
- Simplifies encoder output gathering logic in `gpu_model_runner.py`
- Achieves significant memory savings (up to 12.5x demonstrated in tests)

RESOLVE #25903

## Test Plan
- Unit test added
```bash
pytest tests/multimodal/test_utils.py::test_placeholder_range_get_num_embeds
pytest tests/multimodal/test_utils.py::test_placeholder_range_extract_embeds_range
pytest tests/v1/core/test_encoder_cache_manager.py::test_encoder_cache_with_is_embed_mask
pytest tests/v1/core/test_encoder_cache_manager.py::test_encoder_cache_mask_based_retrieval
pytest tests/v1/worker/test_gpu_model_runner.py::test_encoder_cache_storage_format
```
- regression check
```bash
python examples/offline_inference/vision_language.py --model-type internvl_chat --model internvl_chat --modality image
python examples/offline_inference/vision_language.py --model-type internvl_chat --model qwen2_vl --modality image
```

- inference benchmark

## Test Result
All unit tests passed. Tests validate:

Correct embedding counting with various mask configurations
Proper range extraction for masked positions
Memory savings calculation (12.5x reduction demonstrated)
Cache storage format consistency (raw tensors, not scattered placeholders)

All regression check on relevant models passed

### Inference benchmark
before change
```sh
(EngineCore_DP0 pid=140497) INFO 12-07 07:34:59 [gpu_worker.py:364] Available KV cache memory: 3.93 GiB
(EngineCore_DP0 pid=140497) INFO 12-07 07:35:00 [kv_cache_utils.py:1287] GPU KV cache size: 28,592 tokens
EngineCore_DP0 pid=140497) INFO 12-07 07:35:00 [kv_cache_utils.py:1292] Maximum concurrency for 4,096 tokens per request: 6.98x
```

after change
```sh
(EngineCore_DP0 pid=142442) INFO 12-07 07:38:32 [gpu_worker.py:364] Available KV cache memory: 8.54 GiB
(EngineCore_DP0 pid=142442) INFO 12-07 07:38:33 [kv_cache_utils.py:1287] GPU KV cache size: 62,192 tokens
(EngineCore_DP0 pid=142442) INFO 12-07 07:38:33 [kv_cache_utils.py:1292] Maximum concurrency for 4,096 tokens per request: 15.18x
```

---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [x] The purpose of the PR, such as "Fix issue https://github.com/vllm-project/vllm/issues/25903". 
- [x] The test plan, such as providing test command.
- [x] The test results, such as pasting the results comparison before and after, or e2e results
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model.
- [ ] (Optional) Release notes update. If your change is user facing, please update the release notes draft in the [Google Doc](https://docs.google.com/document/d/1YyVqrgX4gHTtrstbq8oWUImOyPCKSGnJ7xtTpmXzlRs/edit?tab=t.0).
</details>

sunYtokki committed:
[Core] Optimize encoder cache with mask-based storage

This commit optimizes the encoder cache to store only actual embeddings
rather than full placeholder tensors by leveraging is_embed masks. This
achieves memory savings for multimodal inputs.

The optimization reduces memory footprint for vision encoder outputs by
storing only the positions marked as embeddings in the is_embed mask,
rather than storing the full scattered placeholder tensor.

Signed-off-by: Sun Kim <sunytokki@gmail.com>

DarkLight1337 reviewed (commented):
Otherwise the PR looks good. Can you run some evals to verify the correctness? I am not sure whether we have evals for video inputs, so maybe you can just use an image eval on a relevant model like InternVL.

cc @ywang96

ywang96 commented on vllm/v1/worker/gpu_model_runner.py:
```diff
@@ -2245,17 +2240,29 @@ def _gather_mm_embeddings(
+                    # Slice the cached encoder output using the actual embedding indices
+                    mm_embeds_item = encoder_output[
+                        num_embeds_before : num_embeds_before + num_embeds_in_range
+                    ]
```
IMO we should create a cumulative sum of embeddings per position as an attribute of `mm_position`/`PlaceholderRange` instead of computing it at runtime/enginecore, then we can simply do something like

```python
embed_cumsum = pos_info.embed_cumsum # computed from is_embed.to(torch.int32).cumsum(dim=0)
num_embeds_before = int(embed_cumsum[start_idx - 1]) if start_idx > 0 else 0
num_embeds_in_range = int(embed_cumsum[end_idx - 1]) - num_embeds_before 
mm_embeds_item = encoder_output[
    num_embeds_before : num_embeds_before + num_embeds_in_range
]
```
We should also make this code snippet a class method of `PlaceholderRange` instead of having it inside model runner.

ywang96 replied:
> We probably don't want to construct this on P0

Why so?

DarkLight1337 replied:
It introduces additional serialization/deserialization cost. After thinking a bit more, I think we should run some benchmarks to compare the two approaches. Maybe this can be done in another PR?

ywang96 replied:
@ryansherby Hey I actually opened #30475 already for this - see if it works for you

ywang96 reviewed (changes requested):
Thanks for the contribution - I left some comments and please take a look.

Could you also run some benchmarks on models to ensure this does not cause performance regression?

ywang96 commented on vllm/v1/worker/gpu_model_runner.py:
```diff
@@ -2245,17 +2240,29 @@ def _gather_mm_embeddings(
                 req_start_pos = req_start_idx + start_pos - num_computed_tokens
                 is_mm_embed[req_start_pos + start_idx : req_start_pos + end_idx] = (
-                    True if is_embed is None else is_embed
+                    True if is_embed_slice is None else is_embed_slice
```
Why did we need to rename this to `is_embed_slice`?

sunYtokki commented on vllm/multimodal/inputs.py:
```diff
@@ -188,7 +188,9 @@ def extract_embeds_range(self) -> list[tuple[int, int]]:
         if self.is_embed is None:
-            return [(self.offset, self.offset + self.length)]
+            if self.length == 0:
+                return []
```
make sense will address this in the next revision

sunYtokki committed:
clean up unused functions, address cosmetic comments

Signed-off-by: Sun Kim <sunytokki@gmail.com>

sunYtokki commented:
> I just realized this PR still does not align encoder cache budget with the actual space...

I had a seperate PR to align encoder cache bedet https://github.com/vllm-project/vllm/pull/30536

sunYtokki committed:
Fix encoder cache budgeting to use embedding counts for sparse masks

Use actual embedding counts instead of placeholder lengths for encoder
cache allocation and scheduler budgeting. Critical for models with
sparse embedding density (e.g., Qwen3-VL where 100 positions may only
have 8 actual embeddings).

- Update EncoderCacheManager to use embedding counts for eviction
- Update Scheduler to budget encoder tokens using embedding counts
- Add Request.get_num_encoder_tokens() to return actual embedding count

Signed-off-by: Sun Kim <sunytokki@gmail.com>

sunYtokki committed:
restore `scatter_mm_placeholders` and `gather_mm_placeholders` to `vllm.v1.worker.utils`

- These functions are maintained for backward compatibility with external hardware backends
- Core GPU/CPU/XPU backends use the new compact cache format (no functional change)
- TPU backend unaffected (never used these functions)

Signed-off-by: Sun Kim <sunytokki@gmail.com>

sunYtokki commented on vllm/v1/worker/gpu_model_runner.py:
```diff
@@ -166,9 +166,7 @@
     bind_kv_cache,
-    gather_mm_placeholders,
     sanity_check_mm_encoder_outputs,
-    scatter_mm_placeholders,
```
I had to restore those utility functions with a note because XPU and HPU still depend on them, and removing them causes import errors. It looks like this requires coordination with external packages. I’ll open an issue in https://github.com/vllm-project/vllm-gaudi once this change is merged.

sunYtokki committed:
Add deprecation warnings and fix zero-length range handling

Signed-off-by: Sun Kim <sunytokki@gmail.com>

sunYtokki committed:
Use typing_extensions.deprecated for deprecation warning

Signed-off-by: Sun Kim <sunytokki@gmail.com>

sunYtokki commented on vllm/v1/worker/utils.py:
```diff
@@ -215,6 +216,14 @@ def scatter_mm_placeholders(
             tokens need to be filled with multimodal embeddings.
             Shape: `(num_placeholders, num_embeds)`
     """
+    warnings.warn(
```
updated!

sunYtokki committed:
Fix encoder cache optimization with cumulative sum and profiling API refactor

1. **Cumulative Sum Optimization (vllm/multimodal/inputs.py)**:
   - Add embeds_cumsum cached property for O(1) cumulative sum access
   - Add get_embeds_indices_in_range() for efficient embedding index
     lookups
   - Enable O(1) range-to-embedding mapping instead of O(n) counting
   - Add comprehensive unit tests (26 test cases) in test_utils.py

2. **Profiling API Refactor (vllm/multimodal/profiling.py)**:
   - Replace get_mm_max_contiguous_tokens() with get_mm_max_tokens()
   - Remove mm_embeddings_only parameter - always return embedding
     counts
   - Update API to focus on actual embeddings vs placeholder tokens
   - Simplify profiling logic for encoder cache sizing

3. **Test Updates**:
   - Update test_mllama4.py to use new get_mm_max_tokens() API
   - Change assertions to validate embedding counts (total_num_patches)
   - Update registry.py to use refactored profiling API

This builds on previous work (commit 7857ceae5) that updated scheduler
budgeting to use embedding counts. Together, these changes provide:
- Correct memory budgeting for sparse multimodal embeddings
- Efficient range queries for partial scheduling
- Fix for CI multiprocessing initialization errors

Co-authored-by: Roger Wang <rogerw@vllm.ai>

Signed-off-by: Sun Kim <sunytokki@gmail.com>
